### PR TITLE
Pass locals in to the template object on construction

### PR DESCRIPTION
**Pull request by tenderlove**
This commit ensures that locals are passed in to the template objects
when they are constructed, then removes the `locals=` mutator on the
template object.  This means we don't need to mutate Template objects
with locals in the `decorate` method.

Similar to #35408 and #35406

### CODE REVIEW

**Comment by tenderlove** [comment]
I updated this PR to include a commit that deletes the `decorate` method and makes the `Template` object a read-only object.

File: actionview/lib/action_view/template/resolver.rb [hunk 1]
```diff
@@ -165,7 +169,7 @@ def find_all_with_query(query) # :nodoc:
     # This is what child classes implement. No defaults are needed
     # because Resolver guarantees that the arguments are present and
     # normalized.
-    def find_templates(name, prefix, partial, details, outside_app_allowed = false)
+    def find_templates(name, prefix, partial, details, outside_app_allowed = false, locals = [])
       raise NotImplementedError, "Subclasses must implement a find_templates(name, prefix, partial, details, outside_app_allowed = false) method"
```
Comment (georgeclaghorn) [inline actionview/lib/action_view/template/resolver.rb]: Does the method signature in this error message need updating?

File: actionview/lib/action_view/template.rb [hunk 2]
```diff
@@ -122,37 +122,42 @@ def self.finalize_compiled_template_methods=(_)
...
+
       @source            = source
       @identifier        = identifier
       @handler           = handler
       @compiled          = false
-      @original_encoding = nil
-      @locals            = details[:locals] || []
-      @virtual_path      = details[:virtual_path]
+      @locals            = locals
+      @virtual_path      = virtual_path
```
Comment (kaspth) [inline actionview/lib/action_view/template.rb]: How is the virtual_path built here, like it was with Path.build in decorate? I don’t see how decorate can be removed yet.

Comment (tenderlove) [reply actionview/lib/action_view/template.rb]: 1. [`find_templates` uses `Path.build` and passes that to `query`](https://github.com/rails/rails/blob/0e13941dbe9c4840175e5c6f5b2c3f572249a4ec/actionview/lib/action_view/template/resolver.rb#L210-L213)
2. [`query` passes the virtual path to the template constructor](https://github.com/rails/rails/blob/0e13941dbe9c4840175e5c6f5b2c3f572249a4ec/actionview/lib/action_view/template/resolver.rb#L222-L228)

File: actionview/lib/action_view/template/resolver.rb (hunk 1 above)
Comment (tenderlove) [reply actionview/lib/action_view/template/resolver.rb]: ah yep, I'll fix that

File: actionview/lib/action_view/template.rb (hunk 2 above)
Comment (kaspth) [reply actionview/lib/action_view/template.rb]: Ah, thanks! I was looking for the build_path method (which uses Path.build) and found no occurrences, so I got confused. Seems like we could kill that build_path private method. I’ll take a look later unless you want to look into it.

File: actionview/lib/action_view/template/resolver.rb [hunk 3]
```diff
@@ -143,14 +143,18 @@ def clear_cache
 
     # Normalizes the arguments and passes it on to find_templates.
     def find_all(name, prefix = nil, partial = false, details = {}, key = nil, locals = [])
+      locals = locals.map(&:to_s).sort!.freeze
+
       cached(key, [name, prefix, partial], details, locals) do
-        find_templates(name, prefix, partial, details)
+        find_templates(name, prefix, partial, details, locals)
```
Comment (jhawthorn) [inline actionview/lib/action_view/template/resolver.rb]: I think this needs to be

``` suggestion
        find_templates(name, prefix, partial, details, false, locals)
```

File: actionview/lib/action_view/template.rb (hunk 2 above)
Comment (tenderlove) [reply actionview/lib/action_view/template.rb]: You're totally right.  I'll rm that method.  Thank you!

File: actionview/lib/action_view/template.rb [hunk 4]
```diff
@@ -261,11 +266,11 @@ def encode!
     # to ensure that references to the template object can be marshalled as well. This means forgoing
     # the marshalling of the compiler mutex and instantiating that again on unmarshalling.
     def marshal_dump # :nodoc:
-      [ @source, @identifier, @handler, @compiled, @original_encoding, @locals, @virtual_path, @updated_at, @format, @variant ]
+      [ @source, @identifier, @handler, @compiled, @locals, @virtual_path, @updated_at, @format, @variant ]
     end
 
     def marshal_load(array) # :nodoc:
-      @source, @identifier, @handler, @compiled, @original_encoding, @locals, @virtual_path, @updated_at, @format, @variant = *array
+      @source, @identifier, @handler, @compiled, @locals, @virtual_path, @updated_at, @format, @variant = *array
```
Comment (rafaelfranca) [inline actionview/lib/action_view/template.rb]: Should we handle the case were people to marshal load an object created before this change?

**Review by rafaelfranca — APPROVED** [review]

File: actionview/lib/action_view/template.rb (hunk 4 above)
Comment (tenderlove) [reply actionview/lib/action_view/template.rb]: I don't think so.  I think the reason we introduced the marshal stuff is for parallel testing ([according to this comment](https://github.com/rails/rails/blob/082130d1d43620ca29c2f2f169a5059fa0065a8c/actionview/lib/action_view/template.rb#L265-L267).  Those shouldn't stick around.

Comment (rafaelfranca) [reply actionview/lib/action_view/template.rb]: 👍

**Pull request merged**